libminifi/include/core/ProcessGroup.h

```diff
@@ -207,11 +199,18 @@ class ProcessGroup : public CoreComponent {
     return nullptr;
   }
   // findProcessor based on UUID
-  std::shared_ptr<Processor> findProcessorById(const utils::Identifier& uuid, Traverse traverse = Traverse::IncludeChildren) const;
+  Processor* findProcessorById(const utils::Identifier& uuid, Traverse traverse = Traverse::IncludeChildren) const;
   // findProcessor based on name
-  std::shared_ptr<Processor> findProcessorByName(const std::string &processorName, Traverse traverse = Traverse::IncludeChildren) const;
+  Processor* findProcessorByName(const std::string &processorName, Traverse traverse = Traverse::IncludeChildren) const;
+
+  void getProcessorController(const std::string& name, std::vector<state::StateController*>& controllerVec,
+    const std::function<std::unique_ptr<state::ProcessorController>(Processor&)>& controllerFactory);
+
+  void getAllProcessors(std::vector<Processor*>& processor_vec) const;
+
+  void getAllProcessorControllers(std::vector<state::StateController*>& controllerVec,
+    const std::function<std::unique_ptr<state::ProcessorController>(Processor&)>& controllerFactory);
```

Comment:
I don't think `ProcessGroup` should be concerned with the existence of `ProcessorController`s, because it's related to running the flow, not the flow graph itself. I suggest moving back the functionality to `FlowController`, but keeping the functions separated. This doesn't apply to `getAllProcessors`, only to `getProcessorController` and `getAllProcessorControllers`.

Reply:
Done.